experimental/uefi/loader/src/crosvm.rs

```diff
@@ -14,32 +14,88 @@
 // limitations under the License.
 //
 
-use crate::vmm::{Params, Vmm};
+use crate::{
+    vmm::{Params, Vmm},
+    ReadWrite,
+};
 use anyhow::Result;
 use async_trait::async_trait;
 use command_fds::tokio::CommandFdAsyncExt;
 use log::info;
 use std::{
+    io::{Error, ErrorKind, Read, Write},
     net::Shutdown,
-    os::unix::{
-        io::{AsRawFd, OwnedFd},
-        net::UnixStream,
-    },
+    os::unix::{io::AsRawFd, net::UnixStream},
     process::Stdio,
 };
+use vsock::VsockStream;
+
+/// The guest VM context ID for virtio vsock connecitons.
+const VSOCK_GUEST_CID: u32 = 3;
+
+/// The guest VM virtio vsock port.
+const VSOCK_GUEST_PORT: u32 = 1024;
 
 pub struct Crosvm {
     console: UnixStream,
     instance: tokio::process::Child,
 }
 
+/// A wrapper for a vsock stream that can be used as a communication channel between the loader and
+/// the guest VM.
+///
+/// The vsock channel can only be created after the VM is booted, so we start without a stream and
+/// only connect later when we try to use it for the first time.
```

Comment:
Where does "connect later" actually happen: in `ensure_stream`? Do we check every time we write/read? Is this negligible?

Reply:
Yes, it happens in ensure_stream. We do check every time, but it is negligible, as it just checks whether the inner stream is `None` or `Some` to decide whetehr it is connected or not.